content/en/blogs/faas-openfunction.md

```diff
@@ -0,0 +1,300 @@
+            return fmt.Errorf("Function failed to register: %v\n", err)
+        }
+    } else if fnOpenFunction, ok := fn.(func(*ofctx.OpenFunctionContext, []byte) ofctx.RetValue); ok {
+        if err := functionframeworks.RegisterOpenFunction(ctx, fnOpenFunction); err != nil {
+            return fmt.Errorf("Function failed to register: %v\n", err)
+        }
+    } else {
+        err := errors.New("unrecognized function")
+        return fmt.Errorf("Function failed to register: %v\n", err)
+    }
+    return nil
+}
+
+func main() {
+    if err := register(userfunction.HelloWorld); err != nil {
+        log.Fatalf("Failed to register: %v\n", err)
+    }
+
+    if err := functionframeworks.Start(); err != nil {
+        log.Fatalf("Failed to start: %v\n", err)
+    }
+}
+```
+
+Before the application is started, the function is registered as an HTTP function, a CloudEvent function, or an OpenFunction function. When the registration is complete, `functionframeworks.Start` is invoked to start the application.
+
+### Build
+
+Once we have an application, we also need to build it into a container image. Kubernetes has deprecated dockershim and no longer uses Docker as the default container runtime. Thus, we cannot build container images through Docker-in-Docker in a Kubernetes cluster. Is there any other options for building images? How can we manage pipelines for building images?
+
+Tekton is an excellent pipeline tool that originally was a Knative subproject and later donated to [the Continuous Delivery Foundation](https://cd.foundation/). Tekton's pipeline logic is quite intuitive and can be divided into three steps: pulling code, building images, and pushing images. Each step is defined as a Task in Tekton, and all the Tasks form into a pipeline.
+
+There are several tools for building container images, such as Kaniko, Buildah, BuildKit, and Cloud Native Buildpacks (CNB). The first three of them rely on Dockerfiles to build container images, while CNB is the latest technology emerging from the cloud-native field. Sponsored by Pivotal and Heroku, CNB, instead of relying on Dockerfiles, automatically detects the code to be built and generates OCI-compliant container images. This is an astonishing technology that has been adopted by Google Cloud, IBM Cloud, Heroku, and Pivotal. For example, many of the images on Google Cloud are built with CNB.
+
+With these available image building tools, how can we let users select and switch at will between these tools during function building? That entails another project, [Shipwright](https://github.com/shipwright-io/build), an open-source project by Red Hat and IBM. It is dedicated to building container images in a Kubernetes cluster, and is donated to the CD Foundation. With Shipwright, you can flexibly switch between the four image building tools mentioned above as it provides a unified API interface that packages different building methods.
```

Comment:
is donated to => has been donated to?

Reply:
The simple present tense here is a statement of fact. Why use present perfect tense here?